### PR TITLE
Builtin len

### LINKED ISSUE #112: Understand len builtin
Originally reported by: **Claudiu Popa (BitBucket: [PCManticore](http://bitbucket.org/PCManticore), GitHub: @PCManticore)**

---

---
- Bitbucket: https://bitbucket.org/logilab/astroid/issue/112

### PR DESCRIPTION
**Pull request by brycepg**
Implement inference for len builtin 

Close #112

### CODE REVIEW

**Comment by brycepg** [comment]
Something in package_resources might be inferred as len for some reason. (seeing from the lint errors), will checkout out.

File: astroid/helpers.py [hunk 1]
```diff
@@ -199,3 +199,44 @@ def class_instance_as_index(node):
...
     return None
+
+
+def object_len(node, context=None):
+    """Infer length of given node object
+
+    :param Union[nodes.ClassDef, nodes.Instance] node:
+    :param node: Node to infer length of
+
+    :raises AstroidTypeError: If an invalid node is given
```
Comment (PCManticore) [inline astroid/helpers.py]: I can't quite parse this sentence, can you rephrase it?

File: astroid/helpers.py [hunk 2]
```diff
@@ -199,3 +199,44 @@ def class_instance_as_index(node):
...
+        for an invalid type is returned
+    :raises InferenceError: If the given node cannot be inferred
+        or if multiple nodes are inferred
+    :rtype int: Integer length of node
+    """
+    inferred_node = safe_infer(node, context=context)
+    if inferred_node is None or inferred_node is util.Uninferable:
+        raise exceptions.InferenceError(node=node)
+    if inferred_node.qname() in ('builtins.str', 'builtins.bytes'):
```
Comment (PCManticore) [inline astroid/helpers.py]: I'd suggest to check against concrete types, e.g `Dict, List, Tuple` rather than verifying their inner nodes. This way it will be more clear what types are these.

File: astroid/helpers.py [hunk 3]
```diff
@@ -199,3 +199,44 @@ def class_instance_as_index(node):
...
+    try:
+        node_type = object_type(inferred_node, context=context)
+        len_call = next(node_type.igetattr("__len__", context=context))
+    except exceptions.AttributeInferenceError:
+        raise exceptions.AstroidTypeError(
+            "object of type '{}' has no len()"
+            .format(len_call.pytype()))
+
+    try:
+        result_of_len = next(len_call.infer_call_result(node, context))
```
Comment (PCManticore) [inline astroid/helpers.py]: The StopIteration is odd, we should never return it per se, but rather raise an astroid exception. Can you check what exactly is triggering a StopIteration?

File: astroid/tests/unittest_brain.py [hunk 4]
```diff
@@ -1011,5 +1011,144 @@ def _get_result(code):
     return _get_result_node(code).as_string()
 
 
+class TestLenBuiltinInference:
+    def test_len_list(self):
+        # Uses .elts
+        node = astroid.extract_node("""
+        len(['a','b','c'])
+        """)
+        node = next(node.infer())
```
Comment (PCManticore) [inline astroid/tests/unittest_brain.py]: Let's also verify in a test that we return a Const node.

File: astroid/tests/unittest_brain.py [hunk 5]
```diff
@@ -1011,5 +1011,144 @@ def _get_result(code):
...
+
+    def test_len_failure_missing_variable(self):
+        node = astroid.extract_node("""
+        len(a)
+        """)
+        with pytest.raises(astroid.InferenceError):
+            next(node.infer())
+
+    def test_len_bytes(self):
+        """Use .length method to find bytes value"""
```
Comment (PCManticore) [inline astroid/tests/unittest_brain.py]: What's the `.length` method?

File: astroid/helpers.py (hunk 3 above)
Comment (brycepg) [reply astroid/helpers.py]: I'll double check, but I think the StopIteration is coming from classes that don't have a __len__ method

File: astroid/tests/unittest_brain.py (hunk 5 above)
Comment (brycepg) [reply astroid/tests/unittest_brain.py]: I think that was a typo, thanks for the catch

File: astroid/helpers.py (hunk 2 above)
Comment (brycepg) [reply astroid/helpers.py]: More qname?

File: astroid/helpers.py (hunk 1 above)
Comment (brycepg) [reply astroid/helpers.py]: yeah

File: astroid/helpers.py (hunk 3 above)
Comment (brycepg) [reply astroid/helpers.py]: next raises `StopIteration` when the given iterator is empty

Comment (PCManticore) [reply astroid/helpers.py]: What I'm trying to say is that if someone does `next(node.infer_call_result())` that should either fail with InferenceError/return Uninferable or raise some other exception. It shouldn't fail with `StopIteration` explicitly, which usually shows a potential bug. If we put this in some other way, tools that use astroid, for instance pylint, should never care about catching `StopIteration` but only exceptions that astroid provide.

File: astroid/helpers.py (hunk 2 above)
Comment (PCManticore) [reply astroid/helpers.py]: `qname()` could work, as long as it's clear what nodes are we handling here.

File: astroid/helpers.py (hunk 3 above)
Comment (brycepg) [reply astroid/helpers.py]: Oh I see. infer_call_result shouldn't fail silently. I'm thinking it has to do with methods that are defined via the interpreter (written in C)

**Comment by brycepg** [comment]
@PCManticore Heres a reproduction of the StopIteration issue

    import astroid
    code = """
    a = "fast"
    a
    """
    node = astroid.extract_node(code)
    lenmeth = next(node.inferred()[0].igetattr("__len__"))
    next(lenmeth.infer_call_result(None, None))

**Comment by PCManticore** [comment]
Thanks for the reproduction example, Bryce! Given this is not introduced by this PR, there's no reason to be blocked by it. Let's just create an issue with it and it can be investigated in parallel.

File: astroid/helpers.py (hunk 2 above)
Comment (brycepg) [reply astroid/helpers.py]: I think maybe in the future I could transform builtin methods to return correct values instead of being empty. Do you think this is easily do-able?

File: astroid/node_classes.py [hunk 6]
```diff
@@ -2837,6 +2837,21 @@ def catch(self, exceptions): # pylint: disable=redefined-outer-name
         return False
 
 
+class FrozenSet(_BaseContainer):
```
Comment (PCManticore) [inline astroid/node_classes.py]: Hey Bryce, this was actually in objects because there is no frozenset AST node in Python, if I recall correctly. Can you check and revert this change if so?

File: astroid/helpers.py (hunk 2 above)
Comment (PCManticore) [reply astroid/helpers.py]: Sorry I don't understand what you mean :)

File: astroid/node_classes.py (hunk 6 above)
Comment (brycepg) [reply astroid/node_classes.py]: I mean, there's no Instance node in the AST as well

**Comment by brycepg** [comment]
@PCManticore I mean I would like to transform `str.__len__` and make it return `self.value`. Is there a way to transform methods like this?

File: astroid/node_classes.py (hunk 6 above)
Comment (brycepg) [reply astroid/node_classes.py]: Hm but `Instance` is in `Bases`

Comment (brycepg) [reply astroid/node_classes.py]: @PCManticore I could put `FrozenSet` into its own file?

Comment (brycepg) [reply astroid/node_classes.py]: Yeah I'll revert it

**Comment by PCManticore** [comment]
If I understand correctly, you mean to apply this for the builtin methods? What would be the rationale? I'm trying to figure out what you're trying to accomplish. In any case, you could look into the `objectmodel` module, where we do this sort of things, but for this particular use case, I don't quite understand from where the `self.value` is coming from when you infer what `bytes.__len__` is returning, because after all, you already have an instance of `Const` when we detect a `bytes` or `dict` object. If you have a code example, even if it doesn't work, that would be great, since right now I don't quite get your intention with this.

Also, for reference, there is also a `#pylint` channel on IRC where I hang from time to time. If you have any questions that are not necessarily appropriate for an issue, you can also write there and I'll try to respond when I get a chance (might not be completely in sync due to timezone differences)

File: astroid/node_classes.py (hunk 6 above)
Comment (PCManticore) [reply astroid/node_classes.py]: Just wanted to reply to the deleted comments: yes, Instance is in `bases`, but that's because it's needed as a base class for some of the nodes. It's also a bit more generic than what we have in `objects`, since those classes are more specialised versions of an instance of *something*.
Regarding the circular import, in the 2.0 branch that we started some years ago with @ceridwen  there are some changes to alleviate a part of these circular dependencies, although not all of them were addressed. Once I'll finish what I have in store for 3.7 support, I'll focus on bringing those changes to master.

**Comment by brycepg** [comment]
For example, lets say someone's python code calls `str.__len__("f")` or `"f'.__len__()`. Currently this call will fail even with my implementation of the `len` builtin since astroid can't access the C code that implements this length call. I'd like to instead rely astroid providing the machinery to determine the result.

I'd like

```
const = astroid.extract_node("""
a = "f"
a
""").inferred()[0]
bound_method = next(const.igetattr("__len__"))
next(bound_method.infer_call_result(const))
```
`infer_call_result` to yield `nodes.Const(1)` here instead of `StopIteration`

**Comment by PCManticore** [comment]
Gotcha, thanks for the example. It's doable, but not in a simple way. You could take a look at `objectmodel`, where you could implement a custom `__len__` method, that just returns a specialised version per each implementation, which when called, can operate on the argument that was given to.  Here's an example for instance for how we implement the `__get__` method: https://github.com/PyCQA/astroid/blob/8425d915c0186f87da66ad6223588ce70c49031c/astroid/interpreter/objectmodel.py#L280. You might use this one as an inspiration.
